lib/firestore/delete.js

```diff
@@ -280,18 +289,100 @@ FirestoreDelete.progressBar = new ProgressBar("Deleted :current docs (:rate docs
  */
 FirestoreDelete.prototype._recursiveBatchDelete = function() {
   var self = this;
-  return this._getDescendantBatch(this.allDescendants, this.batchSize).then(function(docs) {
-    if (docs.length <= 0) {
-      return Promise.resolve();
+
+  // Tunable deletion parameters
+  var maxOutstanding = 150;
+  var minimumQueueSize = 150;
+
+  // All temporary variables for the deletion queue.
+  var queue = [];
+  var numOutstanding = 0;
+  var pagesRemaining = true;
+  var pageIncoming = false;
+  var lastDocName = undefined;
+
+  var failures = [];
+
+  var queueLoop = function() {
+    if (queue.length == 0 && numOutstanding == 0 && !pagesRemaining) {
+      return true;
     }
 
-    return self._deleteDocuments(docs).then(function(numDeleted) {
-      // Tick the progress bar
-      FirestoreDelete.progressBar.tick(numDeleted);
+    if (queue.length <= minimumQueueSize && pagesRemaining && !pageIncoming) {
+      pageIncoming = true;
 
-      // Recurse to delete another batch
-      return self._recursiveBatchDelete();
-    });
+      self
+        ._getDescendantBatch(self.allDescendants, self.batchSize, lastDocName)
+        .then(function(docs) {
+          pageIncoming = false;
+
+          if (docs.length == 0) {
+            pagesRemaining = false;
+            return;
+          }
+
+          docs.forEach(function(doc) {
+            queue.push(doc);
+          });
+
+          lastDocName = docs[docs.length - 1].name;
+        })
+        .catch(function(e) {
+          logger.debug("Failed to fetch page after " + lastDocName, e);
+          pageIncoming = false;
+        });
+    }
+
+    if (numOutstanding > maxOutstanding) {
+      return false;
+    }
+
+    if (queue.length == 0) {
+      return false;
+    }
+
+    const toDelete = queue.shift();
+    numOutstanding++;
+
+    self
+      ._deleteDocument(toDelete)
+      .then(function() {
+        FirestoreDelete.progressBar.tick(1);
+        numOutstanding--;
+      })
+      .catch(function(e) {
+        // For server errors, just throw the document back into the queue.
+        if (e.status >= 500 && e.status < 600) {
+          logger.debug("Server error deleting doc " + toDelete.name, e);
+          queue.push(toDelete);
+        } else {
+          logger.debug("Fatal error deleting doc " + toDelete.name, e);
+          failures.push(toDelete);
+        }
+
+        numOutstanding--;
+      });
+
+    return false;
+  };
+
+  return new Promise(function(resolve) {
+    var intervalId = setInterval(function() {
+      if (queueLoop()) {
+        clearInterval(intervalId);
+
+        if (failures.length == 0) {
+          resolve();
+        } else {
+          // Fail and indicate in the message which docs
+          // could not be deleted.
+          var failIds = failures.map(function(doc) {
+            return doc.name;
+          });
+          reject("Failed to delete docs: " + failIds);
```

Comment:
you'll have to add `reject` as a parameter when creating the promise in line 369.